### ISSUE TITLE
Migrate from AngularJS to Angular (AngularJS End of Life)

### ISSUE THREAD

**Issue by baldurmen** [glowing-bear/glowing-bear#1078]
As far as I understand, this project depends on AngularJS, which has been End of Life since July 2018: https://www.convective.com/angularjs-end-of-life/

AngularJS is currently in Long Term Support until June 30 2021 and this project should move to an alternative, lilke [Angular](https://angular.io/).

This is likely a large shift, but I think starting now would be good idea :)

**Comment by lorenzhs** [comment]
I don’t think anyone is prepared to put in the required amount of work. Angular is quite different from AngularJS, and the latter won’t stop working. It’s not great, but I don’t think a switch is likely.

**Comment by baldurmen** [comment]
>  and the latter won’t stop working

I respectfully disagree on this. I don't think it's safe or responsible to tell people to use frameworks that have been deprecated and won't receive security updates.

> I don’t think anyone is prepared to put in the required amount of work.

Yeah, that I understand and it's totally OK to say you won't do that work :)

I am not familiar enough with Glowing Bear's code base, but how much work do you estimate that would be? If no contributor is ready to put in that amount of work, maybe we could try to crowdfund it? I'd be happy to give a few bucks if that would help migrate in time for 2021.

FWIW, [I'm currently working on packaging Glowing Bear (not the electron side) in Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=921707) and it certainly won't make it in the next stable release, since AngularJS won't make it either :(

**Comment by lorenzhs** [comment]
> I respectfully disagree on this. I don't think it's safe or responsible to tell people to use frameworks that have been deprecated and won't receive security updates.

Sure, but what are we going to do in the absence of people willing to put in the work? Delete the repository to prevent people from using it?

> I am not familiar enough with Glowing Bear's code base, but how much work do you estimate that would be? If no contributor is ready to put in that amount of work, maybe we could try to crowdfund it? I'd be happy to give a few bucks if that would help migrate in time for 2021.

It's not about money, it's about time and motivation. Instead of porting our codebase, a partial rewrite would probably be the better approach. There's a lot of accumulated cruft (for example around scrolling, connection management, and swipe gestures) that shouldn't be allowed to survive a major restructuring such as porting to a different framework ;)

> FWIW, [I'm currently working on packaging Glowing Bear (not the electron side) in Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=921707) and it certainly won't make it in the next stable release, since AngularJS won't make it either :(

I've argued against packaging Glowing Bear in the past, though mostly on the basis of reasons that don't apply to Debian (#960). I guess that partially stems from us releasing new versions way to infrequently, so that many people use the latest master branch (https://latest.glowing-bear.org) instead. I guess I just don't see the point of packaging a static webpage.

That said, we don't have a hard dependency on the angularjs version, I think the version in Debian should just work.

**Comment by baldurmen** [comment]
> > FWIW, I'm currently working on packaging Glowing Bear (not the electron side) in Debian and it certainly won't make it in the next stable release, since AngularJS won't make it either :(
> 
> I've argued against packaging Glowing Bear in the past, though mostly on the basis of reasons that don't apply to Debian (#960). I guess that partially stems from us releasing new versions way to infrequently, so that many people use the latest master branch (https://latest.glowing-bear.org) instead. I guess I just don't see the point of packaging a static webpage.

If you don't want it to be packaged, I'll respect your decision and I'll stop working on that. Upstream being OK with packaging work is important to me :)

I'm normally not super fond of packaging web apps in Debian because:

1. They are quickly out of date and that poses security issues
2. It's often a PITA

I don't think that really applies to Glowing Bear since it's not updated super often (and doesn't have critical security updates each week) and it's not super hard to package :)

My personal use case for GB being in Debian is I want to be able to use it without using external 3rd party libraries. Being able to use properly vetted dependencies is also a big plus to me. I currently maintain git patches to replace things I dislike and I know a bunch of other Debian users would like to be able to simply "apt install glowing-bear" without going through that hassle too.

> That said, we don't have a hard dependency on the angularjs version, I think the version in Debian should just work.

The angularjs version in Debian has just been bumped to 1.7.9, so that blocker is gone.

**Comment by lorenzhs** [comment]
I'm not against packaging per se, I just don't want it to lead to a worse update experience. If you keep the Debian package reasonably up to date, that's cool. The issues with #960 don't apply here.

How does Debian packaging work for a web app anyway? Distributing the electron runtime for every package seems not-so-smart, so how is it done?

**Comment by Fmstrat** [comment]
We have talked about (in another ticket that I haven't had a chance to work on) using the GitHub Actions Electron Builder for Snapcraft and having GitHub just put the latest release in as a snap for Debian/Ubuntu. 0 maintenance once it's set up.

**Comment by baldurmen** [comment]
> If you keep the Debian package reasonably up to date, that's cool.

That's the goal, yes :)

> How does Debian packaging work for a web app anyway? Distributing the electron runtime for every package seems not-so-smart, so how is it done?

The way Debian does it, packages can't embed dependencies and they have to use packaged ones, that are shared by all other packages. This reduces bloat on systems and makes it possible to keep things updated and secure.

Electron isn't currently packaged in Debian [(although people are working on it)](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=842420), mainly because there are so many fast moving dependencies associated with it.

I don't plan to package the electron part of glowing bear, only the static html.

Eventually, I think it could nice to have a separate package (`glowing-bear-desktop` ?) that implements a `glowing-bear.desktop` entry  [based on the Freedesktop specs](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-1.1.html), so that people can run glowing-bear locally in their preferred browser without using a webserver.

That use case isn't described in the README file so I don't know if you explicitly support it, but it works :p

> We have talked about (in another ticket that I haven't had a chance to work on) using the GitHub Actions Electron Builder for Snapcraft and having GitHub just put the latest release in as a snap for Debian/Ubuntu. 0 maintenance once it's set up.

snap is great for folks who prefer that and I'm sure if you implemented it, lots of folks would use it. A package running on Debian (be it a `.deb` file or some other containerized format) and a package in the main Debian archive are two very different things in term of QA and trust though.

**Comment by AStove** [comment]
How would one go about doing this? Wouldn't it just be a complete rewrite?

**Comment by lorenzhs** [comment]
Pretty much, yeah.

**Comment by cormier** [comment]
I agree with @lorenzhs that this would mean rewriting the entire thing -- with maybe the exception of some backend parts (e.g. the library that implements the weechat relay protocol.)

This is a lot of work and I'm not sure if we'll ever do it, but this is also exciting because it raises a lot of questions an gives us an opportunity to think about things should be done. Here are a few questions that comes off the top of my head.

What library should we use? Should we use Angular because this project has been written with AngularJS or should we use something else? Vue.js? React? How should we choose this libray? Now that this project is mature, we know the specs and our use cases.

When we are set on a library, what guidelines and assumptions should we revisit? Should we add a build step and a distinction between the dev version and the packaged version? Unit tests? Should we keep writing CSS directly or should we use a preprocessor, like Sass?

Finally, if someone is motivated enough to undertake this -- how should we go about building this? Should we just rewrite from scratch? Could we build some sort of unstable hybrid version, with parts written in AngularJS and parts written in the new stack, that people could use right away and work on?

I wish I had more answers than questions but this is all I have for now.

Happy new year to you all!

**Comment by AStove** [comment]
I disagree that the backend parts could be saved. Because that's just where typescript could bring the most advantages, by working with enums, classes and do everything typed. If you want to add something that works on the protocol level in the current implementation it's a snakepit.

One hybrid form I could imagine is keeping the .html roughly the same, treat it as one 'component' in Angular. And then later split it up into seperate components like it should be.

**Comment by cormier** [comment]
> I disagree that the backend parts could be saved. Because that's just where typescript could bring the most advantages, by working with enums, classes and do everything typed. If you want to add something that works on the protocol level in the current implementation it's a snakepit.

Thanks for pointing this out. I made this remark because I do not know Typescript. After you mentionned it, I read about it and you are right, Typescript's static typing could definitely help in making some backend parts easier to maintain.

Moving to Typescript would mean accepting some other design decisions (for example implementing a build process) and could be a big move. However I believe that it should be on the table as well if we want to rigorously discuss all options.

**Comment by AStove** [comment]
https://angular.io/guide/upgrade

There are way to make a hybrid application and migrate piece by piece.

**Comment by lorenzhs** [comment]
That doesn't change anything about any of the other parts, though. A partial upgrade is pointless if it's never going to be finished. Without someone who is prepared to rewrite the whole thing, there's no point.

**Comment by AStove** [comment]
> @angular
> We're told COVID-19 is hurting teams migrating from AngularJS. In response, we are extending the LTS by 6 months. Remember, this only covers security issues and breaking browser bugs.
[link](https://twitter.com/angular/status/1287780634572857357)

Yes, COVID is the reason...
